atef/tests/test_page.py

```diff
@@ -0,0 +1,161 @@
+    qtbot.mouseClick(widget.delete_button, QtCore.Qt.LeftButton)
+
+    assert first_config not in configuration_group_page.data.configs
+
+
+@pytest.mark.parametrize(
+    'group',
+    ['pv_configuration', 'device_configuration', 'tool_configuration']
+)
+def test_add_delete_comparison(
+    request: Any,
+    monkeypatch: Any,
+    qtbot: QtBot,
+    group: AnyDataclass,
+    make_page: Callable,
+):
+    cfg = request.getfixturevalue(group)
+    group_page = make_page(cfg)
+    orig_comp_list = gather_comparisons(cfg)
+
+    qtbot.addWidget(group_page)
+    qtbot.mouseClick(group_page.add_comparison_button, QtCore.Qt.LeftButton)
+    new_comp_list = gather_comparisons(cfg)
+    assert len(new_comp_list) == len(orig_comp_list) + 1
+
+    widget = group_page.comparisons_table.cellWidget(0, 0)
+    deleted_comparison = widget.data
+
+    # mock to auto-confirm deletion
+    monkeypatch.setattr(QtWidgets.QMessageBox, 'question',
+                        lambda *args, **kwargs: QtWidgets.QMessageBox.Yes)
+    qtbot.mouseClick(widget.delete_button, QtCore.Qt.LeftButton)
+
+    final_comp_list = gather_comparisons(cfg)
+    assert deleted_comparison not in final_comp_list
+
+
+@pytest.mark.parametrize(
+    'group',
+    ['pv_configuration', 'device_configuration', 'tool_configuration']
+)
+def test_change_attr(
+    request: Any,
+    qtbot: QtBot,
+    group: AnyDataclass,
+    make_page: Callable,
+):
+    cfg = request.getfixturevalue(group)
+    orig_comps = gather_comparisons(cfg)
+    group_page = make_page(cfg)
+    qtbot.addWidget(group_page)
+
+    row_widget = group_page.comparisons_table.cellWidget(0, 0)
+    new_idx = pick_different_combo_option(row_widget.attr_combo)
+    if not new_idx:
+        return
+
+    row_widget.attr_combo.setCurrentIndex(new_idx)
+    row_widget.attr_combo.activated.emit(new_idx)
+    new_comps = gather_comparisons(cfg)
+    assert len(new_comps) == len(orig_comps)
+    assert new_comps != orig_comps
+
+
+@pytest.mark.parametrize(
+    'group',
+    ['pv_configuration', 'device_configuration', 'tool_configuration']
+)
+def test_change_comparison(
+    request: Any,
+    monkeypatch: Any,
+    qtbot: QtBot,
+    group: AnyDataclass,
+    make_page: Callable
+):
+    cfg = request.getfixturevalue(group)
+    group_page = make_page(cfg)
+    qtbot.addWidget(group_page)
+
+    # get comparison page
+    row_widget = group_page.comparisons_table.cellWidget(0, 0)
+    row_widget.child_button.clicked.emit()
+    comp_page = group_page.full_tree.currentItem().widget
+    old_comp = comp_page.data
+
+    assert isinstance(comp_page, ComparisonPage)
+    new_idx = pick_different_combo_option(comp_page.specific_combo)
+
+    monkeypatch.setattr(QtWidgets.QMessageBox, 'question',
+                        lambda *args, **kwargs: QtWidgets.QMessageBox.Yes)
+    comp_page.specific_combo.setCurrentIndex(new_idx)
+    comp_page.specific_combo.activated.emit(new_idx)
```

Comment:
I'm surprised there isn't an easy way to activate comboboxes via qtbot, it looks like it just gives you mouse and key primitives... oh well

Reply:
Also important to note here is that you have to both change the index and emit the activated signal.  I spent too long figuring that out

Reply:
right, because `activated` implies that a human clicked on it
https://doc.qt.io/qt-6/qcombobox.html#activated

Reply:
this is the sort of thing I'd expect qtbot to be able to do for you 😞